### PR TITLE
Mock netlink and test improvements

### PR DESCRIPTION
**Pull request by fasaxc**
## Description

- Add support for NeighXXX operations in mock netlink. These will replace the shell-out to 'arp' in a future update.
- Police interface index.  Prevents creating a cali interface that overlaps lo.
- Fix up tests accordingly.
- Fix duplicate wireguard test descriptions.
- Fix test leakage in wireguard tests.
- Return a real netlink.LinkNotFound error from the tests; requires an unsafe hack to populate the field.  Improve check in RouteTable to check for the error type as well as the message.

## Related issues/PRs

CORE-9902

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
TBD
```

## Reminder for the reviewer

Make sure that this PR has the correct labels and milestone set.

Every PR needs one `docs-*` label.

- `docs-pr-required`: This change requires a change to the documentation that has not been completed yet.
- `docs-completed`: This change has all necessary documentation completed.
- `docs-not-required`: This change has no user-facing impact and requires no docs.

Every PR needs one `release-note-*` label.

- `release-note-required`: This PR has user-facing changes. Most PRs should have this label.
- `release-note-not-required`: This PR has no user-facing changes.

Other optional labels:

- `cherry-pick-candidate`: This PR should be cherry-picked to an earlier release. For bug fixes only.
- `needs-operator-pr`: This PR is related to install and requires a corresponding change to the operator.

### CODE REVIEW

File: felix/netlinkshim/mocknetlink/netlink.go [hunk 1]
```diff
@@ -76,21 +78,55 @@ var _ netlinkshim.Interface = (*MockNetlinkDataplane)(nil)
...
+
+	// First check that our struct matches the netlink one...
+	nlType := reflect.TypeOf(LinkNotFoundError)
+	ourType := reflect.TypeOf(myLinkNotFoundError{})
+	if nlType.NumField() != ourType.NumField() {
+		panic("netlink.LinkNotFoundError structure appears to have changed (different number of fields)")
+	}
+	for i := 0; i < ourType.NumField(); i++ {
+		nlFieldType := nlType.Field(i).Type
+		outFieldType := ourType.Field(i).Type
```
Comment (nelljerram) [inline felix/netlinkshim/mocknetlink/netlink.go:105]: do you mean "our" instead of "out" ?

File: felix/netlinkshim/mocknetlink/netlink.go [hunk 2]
```diff
@@ -767,6 +831,33 @@ func (d *MockNetlinkDataplane) RouteAdd(route *netlink.Route) error {
...
+	d.mutex.Lock()
+	defer d.mutex.Unlock()
+	defer GinkgoRecover()
+
+	Expect(d.NetlinkOpen).To(BeTrue())
+	if d.shouldFail(FailNextRouteReplace) || d.shouldFail(FailNextRouteAddOrReplace) {
+		return SimulatedError
+	}
+	key := KeyForRoute(route)
+	log.WithField("routeKey", key).Info("Mock dataplane: RouteUpdate called")
```
Comment (nelljerram) [inline felix/netlinkshim/mocknetlink/netlink.go:844]: RouteReplace

**Review by nelljerram — APPROVED** [review]
LGTM.  Two possible nits.

File: felix/netlinkshim/mocknetlink/netlink.go (hunk 1 above)
Comment (fasaxc) [reply felix/netlinkshim/mocknetlink/netlink.go:105]: Ah yes, that's one of my recurring typos :grimacing:

**Review by mazdakn — COMMENTED** [review]
LGTM to my best knowledge. I just have one question. It seems generally speaking we don't check all `nil` values in our test cases and mocks. Isn't it better to check for them and fail cleanly rather than getting panic?

**Comment by fasaxc** [comment]
@mazdakn no particular reason, I suppose I didn't think of passing nil as a particularly high risk bug, so, as long as it does something that fails the test in that case it's OK.

**Pull request merged**